### PR TITLE
add support of seed_everything

### PR DESCRIPTION
**Pull request by Adel-Moumen**
## What does this PR do?

This PR add the support of a new function called `seed_everything` which tries to maximize reproducibility. 

When using two processes it does print: 
```
INFO:speechbrain.utils.seed:[rank: 1] Setting seed to 3403
INFO:speechbrain.utils.seed:[rank: 0] Setting seed to 3402
```
While for one: 
```
INFO:speechbrain.utils.seed:[rank: 0] Setting seed to 3402
```

As a matters of fact, we offset the seed by the rank of the current process.

<details>
  <summary><b>Before submitting</b></summary>

- [ ] Did you read the [contributor guideline](https://speechbrain.readthedocs.io/en/latest/contributing.html)?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing [tests](https://github.com/speechbrain/speechbrain/tree/develop/tests) pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Does your code adhere to project-specific code style and conventions?

</details>

## PR review

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
- [ ] Confirm that the changes adhere to compatibility requirements (e.g., Python version, platform)
- [ ] Review the self-review checklist to ensure the code is ready for review

</details>

### CODE REVIEW

File: speechbrain/utils/seed.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    seed: int = 0, verbose: bool = True, deterministic: bool = False
+) -> int:
+    r"""Function that sets the seed for pseudo-random number generators in: torch, numpy, and Python's random module.
+
+    Arguments
+    ---------
+    seed: the integer value seed for global random state.
+    verbose: Whether to print a message on each rank with the seed being set.
+    deterministic: Whether to set the seed for deterministic operations.
+
```
Comment (pplantinga) [inline speechbrain/utils/seed.py:32]: This docstring is not in the right format, should be:

```
variable: type
  description
```

File: speechbrain/utils/distributed.py [hunk 2]
```diff
@@ -10,12 +10,55 @@
...
+    if rank is not None:
+        return f"[rank: {rank}] {message}"
+    return message
+
+
+def get_rank() -> Optional[int]:
+    r"""Get the rank of the current process.
+
+    This code is taken from the Pytorch Lightning library:
+    https://github.com/Lightning-AI/pytorch-lightning/
```
Comment (pplantinga) [inline speechbrain/utils/distributed.py:44]: Perhaps a permalink to the actual file where it was found could be helpful here

**Review by pplantinga — COMMENTED** [review]
Looks like a good PR that better follows PyTorch recommendations about randomness. I have a few minor comments but it is good enough that it could be merged now.

File: speechbrain/utils/seed.py (hunk 1 above)
Comment (Adel-Moumen) [reply speechbrain/utils/seed.py]: Done.

File: speechbrain/utils/distributed.py [hunk 3]
```diff
@@ -10,12 +10,55 @@
...
 import os
 from functools import wraps
+from typing import Optional
 
 import torch
 
 MAIN_PROC_ONLY: int = 0
 
 
+def rank_prefixed_message(message: str, rank: Optional[int]) -> str:
```
[the remark this answers is not in the document]
Comment (Adel-Moumen) [reply speechbrain/utils/distributed.py]: I think you are right. I left it like that in case an user wants to perform a certain operation if `rank` is `None` but I think its better the way you are proposing.

File: speechbrain/utils/distributed.py (hunk 2 above)
Comment (Adel-Moumen) [reply speechbrain/utils/distributed.py]: Done.

File: docs/experiment.md [hunk 4]
```diff
@@ -94,3 +94,11 @@ We can also read multi-channel SFT data, that will be formatted in this way:
...
 (batch, time_step, n_fft, 2, n_audio_channels)
 ```
+
+## Reproducibility
+
+To improve reproducibility across experiments, SpeechBrain supports its own seeding function located in `speechbrain.utils.seed.seed_everything`. This function sets the seed for various generators such as NumPy, PyTorch, and Python, following the [PyTorch recommendations](https://pytorch.org/docs/stable/notes/randomness.html).
+
+However, due to the differences in how GPU and CPU executions work, results may not be fully reproducible even with identical seeds, especially when training models. This issue primarily affects training experiments.
+
+On the other hand, when preparing data using data preparation scripts, the output of these scripts is independent of the global seeds. This ensures that you will get identical outputs on different setups, even if different seeds are used.
```
[the remark this answers is not in the document]
Comment (Adel-Moumen) [reply docs/experiment.md]: Done. Let me know what you think

**Review by pplantinga — APPROVED** [review]
LGTM!
